File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> indegree(numCourses, 0);
        unordered_map<int, vector<int>> adj;
        vector<int> result;
        
        for(auto vec: prerequisites){
            int u = vec[1];
            int v = vec[0];
            
            indegree[v]++;

            adj[u].push_back(v);
        }

        queue<int> q;

        for(int i=0; i<numCourses; i++){
            if(indegree[i] == 0){
                q.push(i);
                result.push_back(i);
            }
        }
        
        while(!q.empty()){
            int u = q.front();
            q.pop();
            for(auto &v: adj[u]){
                indegree[v]--;
                if(indegree[v] == 0){
                    q.push(v);
                    result.push_back(v);
                }
            }
        }
        
        if(result.size() == numCourses){
            return result;
        }

        return {};
    }
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj_graph = [[] for i in range(numCourses)]
        in_degree = [0] * numCourses
        for i in range(len(prerequisites)):
            dest, source = prerequisites[i]
            adj_graph[source].append(dest)
            in_degree[dest] += 1
        
        courses = []
        q = deque()
        for i in range(numCourses):
            if in_degree[i] == 0:
                q.append(i)
                courses.append(i)
        
        while q:
            current_source = q.pop()
            for edge in adj_graph[current_source]:
                in_degree[edge] -= 1
                if in_degree[edge] == 0:
                    q.append(edge)
                    courses.append(edge)
        
        return courses if len(courses) == numCourses else []